tachyon/crypto/commitments/fri/two_adic_fri_verifier.h

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2022 The Plonky3 Authors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
+// file.
+
+#ifndef TACHYON_CRYPTO_COMMITMENTS_FRI_TWO_ADIC_FRI_VERIFIER_H_
+#define TACHYON_CRYPTO_COMMITMENTS_FRI_TWO_ADIC_FRI_VERIFIER_H_
+
+#include <tuple>
+#include <vector>
+
+#include "tachyon/base/bits.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_config.h"
+#include "tachyon/crypto/commitments/fri/two_adic_fri_proof.h"
+#include "tachyon/math/geometry/dimensions.h"
+#include "tachyon/zk/air/plonky3/challenger/challenger.h"
+
+namespace tachyon::crypto {
+
+template <typename F, typename MMCS>
+struct CommitStep {
+  using Commitment = typename MMCS::Commitment;
+  F beta;
+  Commitment commit;
+  CommitPhaseProofStep<F, MMCS> opening;
+};
+
+template <typename F, typename MMCS>
+F VerifyQuery(size_t index, size_t log_max_height,
+              const TwoAdicFriConfig<MMCS>& config,
+              const std::vector<CommitStep<F, MMCS>>& steps,
+              const std::vector<std::tuple<size_t, F>>& reduced_openings) {
+  F folded_eval = F::Zero();
+  auto ro_iter = reduced_openings.begin();
+  auto ro_end = reduced_openings.end();
+
+  for (size_t step_idx = 0; step_idx < steps.size(); ++step_idx) {
+    size_t log_folded_height = log_max_height - step_idx - 1;
+    if (ro_iter != ro_end && std::get<0>(*ro_iter) == log_folded_height + 1) {
+      folded_eval += std::get<1>(*ro_iter);
+      ++ro_iter;
+    }
+    size_t index_sibling = index ^ 1;
+    size_t index_pair = index >> 1;
+    std::vector<F> evals(2, folded_eval);
+    evals[index_sibling % 2] = steps[step_idx].opening.sibling_value;
+    CHECK(config.mmcs.VerifyOpeningProof(
+        steps[step_idx].commit, {math::Dimensions(2, 1 << log_folded_height)},
+        index_pair, {evals}, steps[step_idx].opening.opening_proof));
+    folded_eval =
+        FoldRow(index_pair, log_folded_height, steps[step_idx].beta, evals);
+    index = index_pair;
+  }
+  CHECK_LT(index, config.Blowup()) << "index was " << index;
+  assert(ro_iter == ro_end);
+
+  return folded_eval;
+}
+
+template <typename F, typename ExtField, typename InputMMCS,
+          typename ChallengeMMCS, typename Challenger, typename Function>
+[[nodiscard]] bool TwoAdicFriPCSVerify(
+    const TwoAdicFriConfig<ChallengeMMCS>& config,
+    const TwoAdicFriProof<ExtField, ChallengeMMCS, F,
+                          std::vector<BatchOpening<F, InputMMCS>>>& proof,
+    Challenger& challenger, Function OpenInput) {
+  size_t cpc_size = proof.commit_phase_commits.size();
+  std::vector<ExtField> betas;
+  betas.reserve(cpc_size);
+  for (size_t i = 0; i < cpc_size; ++i) {
+    challenger.ObserveContainer(proof.commit_phase_commits[i]);
+    betas.emplace_back(challenger.template SampleExtElement<ExtField>());
+  }
+  challenger.ObserveContainer(proof.final_poly.ToBaseFields());
```

Comment:
Please add `begin()` and `end()` in `ExtensionFieldBase` See `tachyon/base/containers/adapters.h`

```suggestion
  challenger.ObserveContainer(proof.final_poly);
```